src/client/datascience/kernel-launcher/jupyterPaths.ts

```diff
@@ -15,29 +15,57 @@ const winJupyterPath = path.join('AppData', 'Roaming', 'jupyter', 'kernels');
 const linuxJupyterPath = path.join('.local', 'share', 'jupyter', 'kernels');
 const macJupyterPath = path.join('Library', 'Jupyter', 'kernels');
 export const baseKernelPath = path.join('share', 'jupyter', 'kernels');
+const CACHE_KEY_FOR_JUPYTER_KERNELSPEC_ROOT_PATH = 'CACHE_KEY_FOR_JUPYTER_KERNELSPEC_ROOT_PATH';
+const CACHE_KEY_FOR_JUPYTER_PATHS = 'CACHE_KEY_FOR_JUPYTER_PATHS';
 
 @injectable()
 export class JupyterPaths {
+    private cachedKernelSpecRootPath?: Promise<string | undefined>;
+    private cachedJupyterPaths?: Promise<string[]>;
     constructor(
         @inject(IPlatformService) private platformService: IPlatformService,
         @inject(IPathUtils) private readonly pathUtils: IPathUtils,
-        @inject(IEnvironmentVariablesProvider) private readonly envVarsProvider: IEnvironmentVariablesProvider
-    ) {}
+        @inject(IEnvironmentVariablesProvider) private readonly envVarsProvider: IEnvironmentVariablesProvider,
+        @inject(IDisposableRegistry) disposables: IDisposableRegistry,
+        @inject(IMemento) @named(GLOBAL_MEMENTO) private readonly globalState: Memento
+    ) {
+        this.envVarsProvider.onDidEnvironmentVariablesChange(
+            () => {
+                this.cachedJupyterPaths = undefined;
+            },
+            this,
+            disposables
+        );
+    }
 
     /**
      * This should return a WRITABLE place that jupyter will look for a kernel as documented
      * here: https://jupyter-client.readthedocs.io/en/stable/kernels.html#kernel-specs
      */
+    @traceDecorators.verbose('Getting Jupyter KernelSpec Root Path')
     public async getKernelSpecRootPath(): Promise<string | undefined> {
```

Comment:
THis logging was very useful, 
On windows this is very slow as `tryGetRealPath` is slow in windows.
Caching takes only 15-20ms and without that it takes around 400ms on windows, every time. this gets called at least 3 times, thus we shaved well over 1s with this simple change.

Reply:
Didn't you have some decorator way to cache results from a function?

Reply:
Oh yes, yo'ure right, let me check

Reply:
The existing decorator will not , because that uses a memory based cache store, i.e. when user reloads VS Code the cache is empty.

I'd need to create a decorator that uses the global memento? Would definitely make the code cleaner & improve readability.
Will do that next week (I.e. will not merge this PR).